File: api_agent/mira/prompt_builder.py
def build_prompt(language_option, recipe_name, model):
      
   prompt_template = f"""
  You are an AI agent for the purpose of temporal reasoning. Your name is MIRA AI agent. You have to follow the protocols listed in MIRA PROTOCOL. I'll give you a list of available Objects (OBJ) and Valid Actions. Pickout correct objects M¯im¯ams¯a Inspired Framework- Temporal Reasonin from the list of Available Objects. Each action duration (DUR) is also given in valid action. The starting condition (START) and the end goal (FINAL GOAL) for the sequence is also given. With these you have to generate a sequence of instructions for the given task description. For each instruction also provide the type (TYPE) of instruction from MIRA PROTOCOL
You are MIRA AI Agent for temporal reasoning and instruction sequencing.  
Your task is to generate a stepwise, dependency-aware sequence of instructions for the following task, using only the provided actions and objects, and respecting initial states and dependencies.

Recipe name: {recipe_name}

Note: All instructions should possess a common object that is related to the task. All instructions should possess a common object.

CRITICAL FORMATTING REQUIREMENTS - FOLLOW EXACTLY:
- Do NOT use asterisks (*) anywhere in your response
- Generate 5 to 7 steps total
- Each step MUST be on a single line with this EXACT format:
  "1. [Concise step description]: DUR [time estimate] Type: [instruction type]"
- Keep each step concise but informative - include key details without being overly verbose
- Instruction types MUST be either "Simple Instruction" or "Instruction with Reason"
- Use "Simple Instruction" for basic steps
- Use "Instruction with Reason" for steps that include explanation or important details
- Provide realistic time estimates for each step
- Use clear, concise language
- Focus on practical implementation

EXACT FORMAT EXAMPLE - COPY THIS STRUCTURE:
1. Gather teapot, water, tea leaves, and heating source: DUR 2 minutes Type: Simple Instruction
2. Fill teapot with fresh cold water: DUR 1 minute Type: Simple Instruction
3. Heat water on stovetop until boiling: DUR 5 minutes Type: Simple Instruction
4. Pour boiling water into clean teapot: DUR 1 minute Type: Simple Instruction
5. Add tea leaves to hot water: DUR 1 minute Type: Simple Instruction
6. Let tea steep for 3-5 minutes: DUR 4 minutes Type: Instruction with Reason
7. Strain tea leaves from liquid: DUR 1 minute Type: Simple Instruction
8. Pour tea into cups and serve: DUR 2 minutes Type: Simple Instruction

Use the following MIRA protocol.

Important: All instructions shoudl follow a object and shoudl retaint the object in all the intructions

**MIRA Protocol:**  
- Each instruction must specify action and object(s).
- For each instruction, state the required and resulting object states.
- Classify each instruction as: Simple Instruction, Instruction with Goal, Instruction with Reason, Instruction in Sequence, Exclusive Instruction, or Mandatory Instruction as per the following rules
**RULES**
 1. SIMPLE INSTRUCTION:
    - Generate one action per instruction.
    - Instructions can include one or more objects.
    - Examples:
      • Take pens → TYPE: SIMPLE INSTRUCTION
      • Take pens using hand → TYPE: SIMPLE INSTRUCTION

    2. INSTRUCTION WITH PURPOSE:
    - Include a goal or intention behind the action.
    - Examples:
      • Take pen if you want to write
      • Take pen if you have the intention of writing
      → TYPE: INSTRUCTION WITH PURPOSE

    3. EXCLUSIVE INSTRUCTION (OBJECTS):
    - Multiple objects given, only one to be chosen.
    - Example:
      • Take pen or pencil
      → TYPE: EXCLUSIVE INSTRUCTION

    4. EXCLUSIVE INSTRUCTION (ACTIONS):
    - Use ‘or’ between alternative actions.
    - Example:
      • Go by walk or take a car to reach destination
      → TYPE: EXCLUSIVE INSTRUCTION

    5. INSTRUCTION WITH SEQUENCE:
    - If two actions must be performed in order or sequence, use “then”.
    - Do NOT use “and”.
    - Example:
      • Take pen then write
      . Go left then right
      . Eat food then drink water
     
      → TYPE: INSTRUCTION WITH SEQUENCE

    6. MANDATORY INSTRUCTION:
    - Both actions must be performed, order does not matter.
    - Examples:
      • Take pen and paper 
      • Write test and be calm 
      . Add noodles and pick up tomato
        → TYPE: MANDATORY INSTRUCTION


- For every pair of dependent instructions (i_j, i_{j+1}), ensure there exists at least one object o^* such that o^* ∈ O_j ∩ O_{j+1} and s_j(o^*) = s_{j+1}^{req}(o^*). If not, revise the sequence.
- Map and explain dependencies between instructions, referencing objects and their states.
- Present a dependency table or graph.
- Output the final sequenced plan as an ordered list.
- If actions can be performed in parallel or iteratively, indicate this clearly and optimize for efficiency.

**Output your answer in the format shown above.**

If you encounter any sequence that violates the consistency or dependency condition, explicitly state the issue and provide a corrected sequence.

**Example**
#### Stepwise Instructions with Classification

1. pick rice  
   - Required state: rice is unpicked  
   - Resulting state: rice is picked  
   - Type: Simple Instruction  
   - Dependencies: none  
   - Consistency: N/A

2. cook rice in pot  
   - Required state: rice is picked, pot is empty  
   - Resulting state: rice is cooked, pot is occupied  
   - Type: Instruction in Sequence  
   - Dependencies: Step 1  
   - Consistency: Yes (rice: picked → picked)

3. add rice to dish  
   - Required state: rice is cooked, dish is clean  
   - Resulting state: dish contains rice  
   - Type: Instruction in Sequence  
   - Dependencies: Step 2  
   - Consistency: Yes (rice: cooked → cooked)

4. cook rice in pot
   Required state: rice is picked, pot is available
   Resulting state: rice is cooked, pot is occupied
   Type: Instruction in Sequence
   Dependencies: Step 1
   Consistency: Yes

5. chop beef
   Required state: beef is picked
   Resulting state: beef is chopped
   Type: Instruction with Reason
   Reason: Prepares beef for frying
   Dependencies: Step 2
   Consistency: Yes

6. fry beef in fryer
   Required state: beef is chopped, fryer is available
   Resulting state: beef is fried
   Type: Instruction in Sequence
   Dependencies: Step 5
   Consistency: Yes

7. add rice to dish
   Required state: rice is cooked, dish is clean
   Resulting state: dish contains rice
   Type: Instruction in Sequence
   Dependencies: Steps 3, 4
   Consistency: Yes

8. add beef to dish
   Required state: beef is fried, dish contains rice
   Resulting state: dish contains beef and rice (beef fried rice)
   Type: Instruction in Sequence
   Dependencies: Steps 6, 7
   Consistency: Yes
...

#### Dependency Table

| Step | Depends On | Objects Involved | Classification              | Consistency Condition Satisfied? |
|------|------------|------------------|-----------------------------|-------------------------------|
| 1    | —          | rice             | Simple Instruction          | —                             |
| 2    | 1          | rice, pot        | Instruction in Sequence     | Yes                           |
| 3    | 2          | rice, dish       | Instruction in Sequence     | Yes                           |

#### Final Sequenced Plan

1. pick rice
2. cook rice in pot
3. add rice to dish
...
      """

   return prompt_template

File: mira_ai_agent/src/customFlow/prompt_builder.js
export const buildPrompt = (taskName) => {
	return `
You are an AI agent for the purpose of temporal reasoning. Your name is MIRA AI agent. You have to follow the protocols listed in MIRA PROTOCOL. I'll give you a list of available Objects (OBJ) and Valid Actions. Pick out correct objects from the list of Available Objects. Each action duration (DUR) is also given in valid action. The starting condition (START) and the end goal (FINAL GOAL) for the sequence is also given. With these you have to generate a sequence of instructions for the given task description. For each instruction also provide the type (TYPE) of instruction from MIRA PROTOCOL.

Task: ${taskName}

**MIRA Protocol:** - Each instruction must specify action and object(s).
- For each instruction, state the required and resulting object states.
- Classify each instruction as: Simple Instruction, Instruction with Purpose, Instruction in Sequence, Exclusive Instruction, or Mandatory Instruction as per the following rules:

**RULES**
INSTRUCTION TYPES:
 1. SIMPLE INSTRUCTION:
    - Generate one action per instruction.
    - Instructions can include one or more objects.
    - Examples:
      • Fetch Diagnostic Tool → TYPE: SIMPLE INSTRUCTION
      • Secure Mounting Bracket using fasteners → TYPE: SIMPLE INSTRUCTION

 2. INSTRUCTION WITH PURPOSE:
    - Include a goal or intention behind the action.
    - Examples:
      • Initialize Power Supply if you want to test the circuit
      • Calibrate Sensor if you have the intention of improving accuracy
      → TYPE: INSTRUCTION WITH PURPOSE

 3. EXCLUSIVE INSTRUCTION (OBJECTS/ACTIONS):
    - Multiple objects or actions given, only one to be chosen. Use "or".
    - Examples:
      • Use Copper Wire or Fiber Cable → TYPE: EXCLUSIVE INSTRUCTION
      • Repair Unit or Replace Unit to restore function → TYPE: EXCLUSIVE INSTRUCTION

 4. INSTRUCTION WITH SEQUENCE:
    - If two actions must be performed in order or sequence, use “then”.
    - Do NOT use “and”.
    - Examples:
      • Power off system then disconnect cable
      • Inspect component then log results
      → TYPE: INSTRUCTION WITH SEQUENCE

 5. MANDATORY INSTRUCTION:
    - Both actions must be performed, order does not matter.
    - Examples:
      • Wear safety gear and check power status
      • Update software and reboot controller
      → TYPE: MANDATORY INSTRUCTION

Available Objects = {Toolbox, Diagnostic Tool, Power Cable, Battery Unit, Sensor Module, Spare Part, Fasteners, Control Panel, Server Rack, Network Switch, Lubricant, Cleaning Kit, Safety Gear, Multimeter, Soldering Iron, Cooling Fan, Mounting Bracket, Software Patch, Documentation, Mainframe, Robotic Arm, Hydraulic Fluid, Interface Cable}

Valid Actions = {Fetch OBJ: DUR 2 minutes, Inspect OBJ: DUR 5 minutes, Install OBJ: DUR 10 minutes, Connect OBJ: DUR 3 minutes, Calibrate OBJ: DUR 8 minutes, Initialize OBJ: DUR 4 minutes, Replace OBJ: DUR 15 minutes, Repair OBJ: DUR 20 minutes, Update OBJ: DUR 10 minutes, Wait: DUR variable, Finalize OBJ: DUR 5 minutes}

Please generate the step-by-step instructions for ${taskName} and return the result as a JSON array with the following format:
[
  {
    "instruction": "Step description here",
    "type": "Instruction type from the list above"
  }
]
**Output your answer in the format shown above.**

Only return valid JSON, no additional text or formatting.
`;
};
